Modules/_datetimemodule.c

```diff
@@ -4839,6 +4852,33 @@ datetime_combine(PyObject *cls, PyObject *args, PyObject *kw)
     return result;
 }
 
+static PyObject *
+_sanitize_isoformat_str(PyObject *dtstr, int *needs_decref) {
+    // `fromisoformat` allows surrogate characters in exactly one position,
+    // the separator; to allow datetime_fromisoformat to make the simplifying
+    // assumption that all valid strings can be encoded in UTF-8, this function
+    // replaces any surrogate character separators with `T`.
+    Py_ssize_t len = PyUnicode_GetLength(dtstr);
+    *needs_decref = 0;
+    if (len <= 10 || !Py_UNICODE_IS_SURROGATE(PyUnicode_READ_CHAR(dtstr, 10))) {
+        return dtstr;
+    }
+
+    PyObject *str_out = PyUnicode_New(len, PyUnicode_MAX_CHAR_VALUE(dtstr));
```

Comment:
`_PyUnicode_Copy()` could be used.

But I'm not sure that arbitrary character (including lone surrogates) should be accepted as a date-time separator. For example 2018-08-24016:10:00 looks pretty confusing.

Reply:
When we developed `datetime.fromisoformat`, the contract of the function was that it should act as the inverse of `datetime.isoformat`, meaning that it should satisfy

```python
datetime.fromisoformat(dt.isoformat(*args, **kwargs)) == dt
```

for all values of `dt`, `args` and `kwargs`. Since `dt.isoformat(sep='\ud800')` and `dt.isoformat(sep='0')` are valid, we need to accept *any* character in the separator position in order to comply with the contract of the function.

It may be worth bringing up the appropriate contract in the `datetime-SIG` mailing list, but it was decided to use a very simple contract so that it's very simple to define what is and is not the correct behavior for this function (which could otherwise grow unwieldy).